File: src/angel/agents/pr_review/state.py
import operator
from typing import Annotated, NamedTuple, TypedDict

from angel.agents._shared import ItemFailure, Review
from angel.clients.github import Comment, PullRequest


class PullRequestContext(NamedTuple):
    pull_request: PullRequest
    comments: list[Comment]


class PrReviewState(TypedDict):
    pull_requests: list[PullRequestContext]
    reviews: list[Review]
    failures: Annotated[list[ItemFailure], operator.add]

File: src/agents/pr-review/state.ts
import { Annotation } from "@langchain/langgraph";
import type { Comment, PullRequest } from "../../clients/github.js";
import type { ItemFailure, Review } from "../shared.js";

/** A pull request due for review, with the comments that decided it. */
export interface PullRequestContext {
  readonly pullRequest: PullRequest;
  readonly comments: Comment[];
}

function replace<T>(_current: T, next: T): T {
  return next;
}

export const PrReviewAnnotation = Annotation.Root({
  pullRequests: Annotation<PullRequestContext[]>({
    reducer: replace,
    default: () => [],
  }),
  reviews: Annotation<Review[]>({ reducer: replace, default: () => [] }),
  /** Accumulates rather than clobbers, because every node writes to it. */
  failures: Annotation<ItemFailure[]>({
    reducer: (current, next) => [...current, ...next],
    default: () => [],
  }),
});

export type PrReviewState = typeof PrReviewAnnotation.State;
